File: haskell/largest-series-product/Series.hs
module Series (digits, slices, largestProduct) where

import Data.Char (digitToInt)

digits :: String -> [Int]
digits = map digitToInt

slices :: Int -> String -> [[Int]]
slices size input = map (\i -> take size $ drop i digits') [0 .. length input - size] where
    digits' = digits input

largestProduct :: Int -> String -> Int
largestProduct _ [] = 1
largestProduct size input = maximum . map product $ slices size input

File: swift/series/Sources/Series/Series.swift
struct Series {
    let digits: [Int]

    init(_ input: String) { 
        digits = input.compactMap(\.wholeNumberValue)
    }
    
    func slices(_ size: Int) -> [[Int]] {
        if size > digits.count { return [] }

        return (0...digits.count - size).map { idx in
            Array(digits[idx..<idx + size])
        }
    }
}
